lib/importer/backend.js

```diff
@@ -0,0 +1,300 @@
+const xlsx = require("node-xlsx").default;
+const crypto = require("crypto");
+const assert = require('node:assert').strict;
+
+// An implementation of the interface described in https://struct.register-dynamics.co.uk/trac/wiki/DataImporter/API
+
+// FIXME: State storage all in-memory for now...
+
+let sessionStore = new Map();
+let jobStore = new Map();
+
+function makeAccessToken(kind) {
+    return kind + "-" + crypto.randomBytes(16).toString("hex");
+}
+
+///
+/// Control Interface
+///
+
+// Create a new import session; returns session ID (sid)
+exports.CreateSession = () => {
+    let sid = makeAccessToken("s");
+    sessionStore[sid] = new Map();
+    return sid;
+};
+
+///
+/// Session Interface
+///
+
+// Set the filename of the file for an import session.
+exports.SessionSetFile = (sid, filename) => {
+    const data = xlsx.parse(filename);
+    let session = {
+        filename: filename,
+        data: data,
+        jobs: new Set(),
+        sheets: new Map(),
+    }
+    data.forEach((sheet) => {
+        session.sheets[sheet.name] = sheet.data;
+    });
+    sessionStore[sid] = session;
+};
+
+function max(x,y) {
+    if(x > y) return x;
+    else return y;
+}
```

Comment:
We could probably use `Math.max(x, y)` as I think it's available with node.

Reply:
Ahah!